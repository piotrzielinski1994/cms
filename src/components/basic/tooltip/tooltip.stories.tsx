import { DEFAULT_VALUE, getFallback } from '@/config/storybook/utils';
import { cn } from '@/utils/tailwind';
import type { Meta, StoryObj } from '@storybook/react';
import { HelpCircle } from 'lucide-react';
import { useTranslations } from 'next-intl';
import { type ComponentProps } from 'react';
import { Tooltip as TooltipComponent } from './tooltip';

type Args = ComponentProps<typeof TooltipComponent> & {
  location: 'left' | 'center' | 'right';
};

const meta = {
  component: TooltipComponent,
  title: 'Components/Basic/Tooltip',
  argTypes: {
    content: { control: 'text' },
    children: { control: 'text' },
    location: {
      control: 'select',
      options: ['left', 'center', 'right'] satisfies Array<Args['location']>,
    },
  },
  args: {
    content: DEFAULT_VALUE,
    children: DEFAULT_VALUE,
    location: 'left',
  },
} satisfies Meta<Args>;

const Render = ({ content, children, location, ...args }: Args) => {
  const t = useTranslations('storybook.basic');
  const classPerLocation = {
    left: 'justify-start',
    center: 'justify-center',
    right: 'justify-end',
  } satisfies Record<Args['location'], string>;

  return (
    <div className={cn('grid', classPerLocation[location])}>
      <TooltipComponent {...args} content={getFallback(content, t('tooltip'))}>
        {getFallback(children, <HelpCircle />)}
      </TooltipComponent>
    </div>
  );
};

const Tooltip: StoryObj<typeof TooltipComponent> = { render: Render };

export { Tooltip };
export default meta;
