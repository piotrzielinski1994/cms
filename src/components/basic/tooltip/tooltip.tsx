import { cn } from '@/utils/tailwind';
import { ComponentPropsWithoutRef, useId, useState } from 'react';

type TooltipProps = ComponentPropsWithoutRef<'button'> & {
  content: string;
};

const Tooltip = ({ content, children, ...props }: TooltipProps) => {
  const [isVisible, setIsVisible] = useState(false);
  const id = useId();

  return (
    <div
      className="relative inline-flex"
      onMouseEnter={() => setIsVisible(true)}
      onMouseLeave={() => setIsVisible(false)}
    >
      <button
        {...props}
        type="button"
        aria-describedby={isVisible ? id : undefined}
        onClick={() => setIsVisible((prev) => !prev)}
        onBlur={() => setIsVisible(false)}
      >
        {children}
      </button>

      {isVisible && (
        <div className="absolute top-full left-1/2 -translate-x-1/2 pt-2">
          <div
            id={id}
            role="tooltip"
            className={cn('py-1 px-2 text-xs bg-foreground text-background')}
          >
            {content}
          </div>
        </div>
      )}
    </div>
  );
};

export { Tooltip };
