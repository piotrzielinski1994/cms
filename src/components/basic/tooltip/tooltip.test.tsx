import { render, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { ComponentProps } from 'react';
import { describe, expect, it } from 'vitest';
import { axe } from 'vitest-axe';
import { Tooltip } from './tooltip';

describe('Tooltip', () => {
  const defaultProps = {
    content: 'Tooltip text',
    children: <span>Trigger</span>,
  } satisfies ComponentProps<typeof Tooltip>;

  it('should have no accessibility violations', async () => {
    const { container, getByRole } = render(<Tooltip {...defaultProps} />);
    await userEvent.hover(getByRole('button'));

    const results = await axe(container);

    expect(results).toHaveNoViolations();
  });

  it('should match the snapshot', async () => {
    const { container, getByRole } = render(<Tooltip {...defaultProps} />);
    await userEvent.hover(getByRole('button'));
    expect(container).toMatchSnapshot();
  });

  it('should show tooltip on hover and hide on mouse leave', async () => {
    const { getByRole, queryByRole } = render(<Tooltip {...defaultProps} />);
    const button = getByRole('button');

    await userEvent.hover(button);
    expect(getByRole('tooltip')).toBeVisible();

    await userEvent.unhover(button);
    expect(queryByRole('tooltip')).toBeNull();
  });

  it('should toggle tooltip on click', async () => {
    const { getByRole, queryByRole } = render(<Tooltip {...defaultProps} />);
    const button = getByRole('button');

    await userEvent.hover(button);
    expect(getByRole('tooltip')).toBeVisible();

    await userEvent.click(button);
    expect(queryByRole('tooltip')).toBeNull();

    await waitFor(() => button.click()); // userEvent approach seems to conflict with hover effect
    expect(getByRole('tooltip')).toBeVisible();
  });
});
