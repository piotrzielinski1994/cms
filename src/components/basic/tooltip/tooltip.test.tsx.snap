// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Tooltip > should match the snapshot 1`] = `
<div>
  <div
    class="relative inline-flex"
  >
    <button
      aria-describedby="«r1»"
      type="button"
    >
      <span>
        Trigger
      </span>
    </button>
    <div
      class="absolute top-full left-1/2 -translate-x-1/2 pt-2"
    >
      <div
        class="py-1 px-2 text-xs bg-foreground text-background"
        id="«r1»"
        role="tooltip"
      >
        Tooltip text
      </div>
    </div>
  </div>
</div>
`;
